### PR TITLE
xapi: add xapi functions.

### PR DESCRIPTION
**Pull request by coocood**
Some functions are not used, but the following PR will use them.
Tests will be added in following PR.

### CODE REVIEW

File: kv/key.go [hunk 1]
```diff
@@ -26,6 +26,29 @@ func (k Key) Next() Key {
 	return buf
 }
 
+// PartialNext returns the next partial key.
+// For example, a key composed with two columns.
+// Next will return a key with the same first column value,
+// but PartialNext will return a key with different first column value.
+// key encoding method must ensure the next different value has the
+// same length as the original value.
+func (k Key) PartialNext() Key {
```
Comment (shenli) [inline kv/key.go]: Next row key?
Any example or unit test to explain this function?

File: xapi/tablecodec/tablecodec.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+		values[i], err = unflatten(values[i], cols[i])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return values, nil
+}
+
+// unflatten converts a raw datum to a column datum.
+func unflatten(datum types.Datum, tp *tipb.ColumnInfo) (types.Datum, error) {
```
Comment (shenli) [inline xapi/tablecodec/tablecodec.go]: Can we use the function in tables package?

File: kv/key.go (hunk 1 above)
Comment (coocood) [inline kv/key.go]: It doesn't have to be row key, it can be next index value.

File: xapi/xapi.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tidb/xapi/tablecodec"
+	"github.com/pingcap/tidb/xapi/tipb"
+)
+
+// SelectResult is used to get response rows from SelectRequest.
+type SelectResult struct {
+	req      *tipb.SelectRequest
```
Comment (shenli) [inline xapi/xapi.go]: Why put request here?

File: xapi/tablecodec/tablecodec.go (hunk 2 above)
Comment (coocood) [inline xapi/tablecodec/tablecodec.go]: `tables` package uses `model.ColumnInfo` type, that requires a convertion.

File: xapi/xapi.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	resp     kv.Response
+	subResp  *tipb.SelectResponse
+	cursor   int
+	finished bool
+}
+
+// Next returns the next row.
+func (r *SelectResult) Next() (handle int64, data []types.Datum, err error) {
```
Comment (shenli) [inline xapi/xapi.go]: Will there be any other interface for executor to access response from multiple region servers?

File: xapi/tablecodec/tablecodec.go (hunk 2 above)
Comment (shenli) [inline xapi/tablecodec/tablecodec.go]: It only depends on FieldType, not ColumnInfo.

File: xapi/xapi.go (hunk 3 above)
Comment (coocood) [inline xapi/xapi.go]: We need column info in req to decode value.

File: xapi/tablecodec/tablecodec.go (hunk 2 above)
Comment (coocood) [inline xapi/tablecodec/tablecodec.go]: FeildType requires conversion too.

**Comment by coocood** [comment]
@shenli PTAL

File: xapi/xapi.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+// Close closes SelectResult.
+func (r *SelectResult) Close() error {
+	return r.resp.Close()
+}
+
+// SubResult represents a subset of select result.
+type SubResult struct {
+	columns []*tipb.ColumnInfo
```
Comment (shenli) [inline xapi/xapi.go]: For "select a+b from t;", there is no physical column. How about use a struct like ResultField?

File: kv/key.go (hunk 1 above)
Comment (shenli) [inline kv/key.go]: OK. Please add an unit test.

File: xapi/tablecodec/tablecodec.go [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+	}
+
+	if !key.HasPrefix(recordPrefixSep) {
+		return 0, errors.Errorf("invalid record key - %q", k)
+	}
+
+	key = key[len(recordPrefixSep):]
+
+	key, handle, err = codec.DecodeInt(key)
+	if err != nil {
```
Comment (zimulala) [inline xapi/tablecodec/tablecodec.go]: Here can return directly?

File: xapi/tablecodec/tablecodec.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+	case mysql.TypeBit:
+		bit := mysql.Bit{Value: datum.GetUint64(), Width: ft.Flen}
+		datum.SetValue(bit)
+		return datum, nil
+	}
+	log.Error(ft.Tp, datum)
+	return datum, nil
+}
+
+// EncodeIndexSeekKey encodes a index value to kv.Key.
```
Comment (shenli) [inline xapi/tablecodec/tablecodec.go]: a -> an

File: xapi/tablecodec/tablecodec.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+			return datum, errors.Trace(err)
+		}
+		datum.SetValue(set)
+		return datum, nil
+	case mysql.TypeBit:
+		bit := mysql.Bit{Value: datum.GetUint64(), Width: ft.Flen}
+		datum.SetValue(bit)
+		return datum, nil
+	}
+	log.Error(ft.Tp, datum)
```
Comment (zimulala) [inline xapi/tablecodec/tablecodec.go]: Remove this line?

File: xapi/tablecodec/tablecodec.go (hunk 7 above)
Comment (zimulala) [inline xapi/tablecodec/tablecodec.go]: a -> an ?

File: xapi/xapi.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	return
+}
+
+// Close closes SelectResult.
+func (r *SelectResult) Close() error {
+	return r.resp.Close()
+}
+
+// SubResult represents a subset of select result.
```
Comment (shenli) [inline xapi/xapi.go]: "SubResult represents a select result from a single region server."?

**Comment by shenli** [comment]
Rest LGTM

File: xapi/xapi.go (hunk 9 above)
Comment (coocood) [inline xapi/xapi.go]: We don't have the concept of region server here.

File: xapi/tablecodec/tablecodec_test.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	key := EncodeRowKey(1, codec.EncodeInt(nil, 2))
+	h, err := DecodeRowKey(key)
+	c.Assert(err, IsNil)
+	c.Assert(h, Equals, int64(2))
+
+	key = EncodeColumnKey(1, codec.EncodeInt(nil, 2), 3)
+	h, err = DecodeRowKey(key)
+	c.Assert(err, IsNil)
+	c.Assert(h, Equals, int64(2))
+
```
Comment (zimulala) [inline xapi/tablecodec/tablecodec_test.go]: Remove this line.

Comment (coocood) [inline xapi/tablecodec/tablecodec_test.go]: This is not finished yet, more tests will be added here.

**Comment by coocood** [comment]
@zimulala PTAL

File: xapi/xapi.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+	subResult = &SubResult{
+		fields: r.fields,
+		reader: reader,
+	}
+	return
+}
+
+// Close closes SelectResult.
+func (r *SelectResult) Close() error {
+	return r.resp.Close()
```
Comment (zimulala) [inline xapi/xapi.go]: need `if r.resp == nil` ?

File: xapi/xapi.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+// SelectResult is used to get response rows from SelectRequest.
+type SelectResult struct {
+	fields []*types.FieldType
+	resp   kv.Response
+}
+
+// Next returns the next row.
+func (r *SelectResult) Next() (subResult *SubResult, err error) {
+	var reader io.ReadCloser
+	reader, err = r.resp.Next()
```
Comment (zimulala) [inline xapi/xapi.go]: need `if r.resp == nil` ?

File: xapi/tablecodec/tablecodec.go [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+		}
+		columnCount--
+	}
+	handle = b
+	return key[prefixLen : len(key)-len(handle)], handle, nil
+}
+
+// record prefix is "t[tableID]_r"
+func appendTableRecordPrefix(buf []byte, tableID int64) []byte {
+	buf = append(buf, tablePrefix...)
```
Comment (zimulala) [inline xapi/tablecodec/tablecodec.go]: Add . at the end of a line?

File: xapi/tablecodec/tablecodec.go [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+	buf = codec.EncodeInt(buf, tableID)
+	buf = append(buf, recordPrefixSep...)
+	return buf
+}
+
+// index prefix is "t[tableID]_i"
```
Comment (zimulala) [inline xapi/tablecodec/tablecodec.go]: Add . at the end of a line?

File: xapi/tablecodec/tablecodec.go [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		columnCount--
+	}
+	handle = b
+	return key[prefixLen : len(key)-len(handle)], handle, nil
+}
+
+// record prefix is "t[tableID]_r"
```
Comment (zimulala) [inline xapi/tablecodec/tablecodec.go]: Add . at the end of a line?

File: xapi/xapi.go (hunk 12 above)
Comment (coocood) [inline xapi/xapi.go]: It will never be nil.

File: xapi/xapi.go [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+		columns = req.IndexInfo.Columns
+	}
+	fields := tablecodec.ProtoColumnsToFieldTypes(columns)
+	return &SelectResult{fields: fields, resp: resp}, nil
+}
+
+// Convert tipb.Request to kv.Request.
+func composeRequest(req *tipb.SelectRequest, concurrency int) (*kv.Request, error) {
+	var tid int64
+	var reqType int64
```
Comment (zimulala) [inline xapi/xapi.go]: Replace `var reqType int64` with `kvReq := &kv.Request{Concurrency: concurrency}` to make the next code more simple ?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**